src/toniarts/openkeeper/world/room/Temple.java

```diff
@@ -102,19 +102,18 @@ protected BatchNode constructFloor() {
             int i = count / waterArea[0].length;
             int j = count % waterArea[0].length;
 
-//            if(waterArea[i][j]) {
-//                Terrain terrainAtPoint = roomInstance.getTerrainAtPoint(p);
-//                terrainAtPoint.getFlags().add(Terrain.TerrainFlag.WATER);
-//                EntityInstance<Terrain> entityInstance = new EntityInstance<>(terrainAtPoint);
-//
-//                waterBatches.add(entityInstance);
-//            }
+            if(waterArea[i][j]) {
+                EntityInstance<Terrain> ent = new EntityInstance<>(getWorldState().getGameState().getLevelData().getMap().getWater());
```

Comment:
This is a bit wrong idea-wise, EntityInstance is like a one continuous area of <T>. I don't remember anymore and I'm too lazy to look at the code at work. But normally we batch the map, 8x8 batches, rooms are batched separately. So is water areas (the surface). Water is batched a bit differently, it is homemade formula, we just create it as one mesh, which is quicker anyway. If you supply the water constructor with many many instances, it might batch them separately (here producing 1 tile patches, which make no sense), or it might work, I don't remember, but ideawise it is a bit wrong.

Batching is essential to performance, something that some of us dislike, but otherwise the object count in the scene would skyrocket. And would be very slow, even though the vertice count is small. Batching unites meshes with the same material to one single mesh.

So here, you would only have one instance, you can even init the intances arraylist capacity to 1. Create the entity. In the loop you would just add the coordinates to the one entity to you have. Thus you would end up with one water entity that spans an area. That is also less objects created and less functions calls :) Less carbage to collect and superior performance :)